polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -2616,91 +2633,215 @@ ValueCategory MLIRScanner::EmitBinRem(const BinOpInfo &Info) {
   }
 }
 
+/// Casts index of subindex operation conditionally.
+static Optional<Value> castSubIndexOpIndex(OpBuilder &Builder, Location Loc,
+                                           ValueCategory Pointer,
+                                           ValueRange IdxList, bool IsSigned) {
+  if (Pointer.val.getType().isa<MemRefType>()) {
+    assert(IdxList.size() == 1 && "SubIndexOP accepts just an index");
+    return ValueCategory(IdxList.front(), false)
+        .IntCast(Builder, Loc, Builder.getIndexType(), IsSigned)
+        .val;
+  }
+  return llvm::None;
+}
+
+ValueCategory MLIRScanner::EmitCheckedInBoundsGEP(mlir::Type ElemTy,
+                                                  ValueCategory Pointer,
+                                                  ValueRange IdxList,
+                                                  bool IsSigned, bool) {
+  assert(mlirclang::isPointerOrMemRefTy(Pointer.val.getType()) &&
+         "Expecting pointer or MemRef");
+  assert(std::all_of(IdxList.begin(), IdxList.end(),
+                     [](mlir::Value Val) {
+                       return Val.getType().isa<IntegerType>();
+                     }) &&
+         "Expecting indices list");
+
+  if (Optional<Value> NewValue =
+          castSubIndexOpIndex(builder, loc, Pointer, IdxList, IsSigned))
+    IdxList = *NewValue;
+
+  return Pointer.InBoundsGEPOrSubIndex(builder, loc, ElemTy, IdxList);
+}
+
+ValueCategory MLIRScanner::EmitPointerArithmetic(const BinOpInfo &Info) {
+  const auto *Expr = cast<BinaryOperator>(Info.getExpr());
+
+  ValueCategory Pointer = Info.getLHS();
+  auto *PointerOperand = Expr->getLHS();
+  ValueCategory Index = Info.getRHS();
+  auto *IndexOperand = Expr->getRHS();
+
+  const auto Opcode = Info.getOpcode();
+  const auto IsSubtraction =
+      Opcode == clang::BO_Sub || Opcode == clang::BO_SubAssign;
+
+  // In a subtraction, the LHS is always the pointer.
+  if (!IsSubtraction &&
+      !mlirclang::isPointerOrMemRefTy(Pointer.val.getType())) {
```

Comment:
```suggestion
  assert(IsSubtraction && mlirclang::isPointerOrMemRefTy(Pointer.val.getType()) && "The LHS is always the pointer in a subtraction");
  if (!mlirclang::isPointerOrMemRefTy(Pointer.val.getType())) {
```

Reply:
Added slightly modified version of this assert.